apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/DroppedSpanStats.java

```diff
@@ -0,0 +1,171 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.impl.transaction;
+
+import co.elastic.apm.agent.objectpool.Allocator;
+import co.elastic.apm.agent.objectpool.ObjectPool;
+import co.elastic.apm.agent.objectpool.Recyclable;
+import co.elastic.apm.agent.objectpool.impl.QueueBasedObjectPool;
+import org.jctools.queues.atomic.MpmcAtomicArrayQueue;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class DroppedSpanStats implements Iterable<Map.Entry<DroppedSpanStats.StatsKey, DroppedSpanStats.Stats>>, Recyclable {
+
+    public static class StatsKey implements Recyclable {
+        private String destinationServiceResource;
+        private Outcome outcome;
+
+        public StatsKey() {
+
+        }
+
+        public StatsKey init(CharSequence destinationServiceResource, Outcome outcome) {
+            this.destinationServiceResource = destinationServiceResource.toString();
+            this.outcome = outcome;
+            return this;
+        }
+
+        public String getDestinationServiceResource() {
+            return destinationServiceResource;
+        }
+
+        public Outcome getOutcome() {
+            return outcome;
+        }
+
+        @Override
+        public void resetState() {
+            destinationServiceResource = null;
+            outcome = null;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            StatsKey statsKey = (StatsKey) o;
+            return destinationServiceResource.equals(statsKey.destinationServiceResource) && outcome == statsKey.outcome;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(destinationServiceResource, outcome);
+        }
+    }
+
+    public static class Stats implements Recyclable {
+        private final AtomicInteger count = new AtomicInteger(0);
+        private final AtomicLong sum = new AtomicLong(0L);
+
+        public int getCount() {
+            return count.get();
+        }
+
+        public long getSum() {
+            return sum.get();
+        }
+
+        @Override
+        public void resetState() {
+            count.set(0);
+            sum.set(0L);
+        }
+    }
+
+    private static final ObjectPool<StatsKey> statsKeyObjectPool = QueueBasedObjectPool.<StatsKey>ofRecyclable(new MpmcAtomicArrayQueue<StatsKey>(512), false, new Allocator<StatsKey>() {
```

Comment:
is 512 a reasonable default?

Reply:
I think that the majority of applications will be using the default values and not very large number of spans in transactions. So 512 is fine for these. A minority will drop a lot and might need a larger value for optimum recycling. If we get complaints in future about too much garbage with a high level of dropped spans, we might need to expose the value as config, but for now this is fine